lib/ecto/soft_delete_repo.ex

```diff
@@ -41,35 +53,42 @@ defmodule Ecto.SoftDelete.Repo do
         {:error, changeset} -> # Something went wrong
       end
 
+      # With schema prefix for multi-tenant databases
+      MyRepo.soft_delete(post, prefix: "tenant_abc")
+
   """
-  @callback soft_delete(struct_or_changeset :: Ecto.Schema.t() | Ecto.Changeset.t()) ::
-              {:ok, Ecto.Schema.t()} | {:error, Ecto.Changeset.t()}
+  @callback soft_delete(
+              struct_or_changeset :: Ecto.Schema.t() | Ecto.Changeset.t(),
+              opts :: Keyword.t()
+            ) :: {:ok, Ecto.Schema.t()} | {:error, Ecto.Changeset.t()}
 
   @doc """
-  Same as `c:soft_delete/1` but returns the struct or raises if the changeset is invalid.
+  Same as `c:soft_delete/2` but returns the struct or raises if the changeset is invalid.
   """
-  @callback soft_delete!(struct_or_changeset :: Ecto.Schema.t() | Ecto.Changeset.t()) ::
-              Ecto.Schema.t()
+  @callback soft_delete!(
+              struct_or_changeset :: Ecto.Schema.t() | Ecto.Changeset.t(),
+              opts :: Keyword.t()
+            ) :: Ecto.Schema.t()
 
   defmacro __using__(_opts) do
     quote do
       import Ecto.Query
       import Ecto.SoftDelete.Query
 
-      def soft_delete_all(queryable) do
-        update_all(queryable, set: [deleted_at: DateTime.utc_now()])
+      def soft_delete_all(queryable, opts \\ []) do
+        set_expr = [set: [deleted_at: DateTime.utc_now()]]
+        update_all(queryable, set_expr, opts)
       end
 
-      def soft_delete(struct_or_changeset) do
-        struct_or_changeset
-        |> Ecto.Changeset.change(deleted_at: DateTime.utc_now())
-        |> update()
+      # Define soft_delete and soft_delete! with options
+      def soft_delete(struct_or_changeset, opts \\ []) do
+        changeset = Ecto.Changeset.change(struct_or_changeset, deleted_at: DateTime.utc_now())
+        __MODULE__.update(changeset, opts)
```

Comment:
If you really wanted, you could probably do something like:

```suggestion
      def soft_delete(struct_or_changeset, opts \\ []) do
        struct_or_changeset
        |> Ecto.Changeset.change(deleted_at: DateTime.utc_now())
        |> __MODULE__.update(opts)
```
and keep those nice pipes!

Reply:
You are right! There is no need to assign that intermediate variable. Using a pipeline is a way cleaner! Also I should stick with the style of the rest of the repo. Making this change now. Thanks for all your input and help on this!

Reply:
no worries, my pleasure!